CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+
+**Software Design Constraints**
+
+- Global data must be documented in the design.
+- Global data must be minimized.
+- Bi-directional dependencies between software elements must be justified.
+- Exceptions and assertions used for control flow must be justified. Exception:
+  a timer exception used for task scheduling.
+- The use of self-modifying software is prohibited.
+
+**Software Design Guidelines**
+
+- The architecture should support one or more levels of abstraction with clearly
+  defined layers.
+- The architecture and design detail should be modular.
+- The design should reduce software complexity.
+- The design should promote reuse.
+- The design should facilitate extensibility.
+- The design should enhance maintainability.
+- The design should result in deterministic software with correct and
+  predictable behavior.
+- Justification will be required when considering software constructs or schemes
+  such as scheduling, interrupt and event-driven architectures, dynamic tasking,
+  re-entry, global data and exception handling.
+- The design should have high cohesion within software elements.
+- The design should have loose coupling between software elements.
+- The design should include well-defined interface descriptions between software
+  elements.
+- The design should include a hardware abstraction layer (HAL) that provides a
+  device-independent interface to hardware services.
+
+**Object-Oriented Software Design Guidelines**
+
+- Where appropriate, make use of proven design patterns. If a design pattern
+  adds complexity, make sure the benefit outweighs the added complexity.
+- Use of interfaces as a means to decrease coupling is recommended; allow
+  clients to depend on interfaces to software elements, not on their
+  implementations.
+- Composition/aggregation is recommended instead of implementation inheritance.
+- Use of inheritance solely to provision for unlikely future product
+  improvements should be avoided. Abstracting an interface and separating it
+  from an implementation as reasonable accommodation for change may be
+  appropriate.
+- Avoid allocating too many responsibilities to a single class.
+
+
+Additional Guidelines and Conventions
+================================================================================
+
+This section contains recommendations, not project standards.
+
+The guidelines should be considered during code review and deviations in the
+code base should be justified or resolved as soon as possible.
+
+**File Naming**
+
+C++ code should adhere to these naming conventions:
+
+  - All source code filenames should be lower case.
+
+  - C++ files should use the .cpp and .h extensions (.hpp extension
+    should not be used).
+
+  - Header files that are #included in the body of another file (e.g. to define
+    a reusable lookup table) should use the -inline.h suffix.
+
+  - All C++ files associated with a design package should be prefixed with the
```

Comment:
We should rename some of our packages to be briefer if we are sticking by this. I have some massively long package names

Reply:
I think it's definitely an optional standard.  Our packages were `snsr`, `nav`, `apex`, `flpn`, etc.

Reply:
Either way.  Our build system sort of ignored our file structure, so we would write `#include "snsr_subsystem.h" instead of `#include "system/snsr/subsystem.h", even though that's where that file actually existed, so we needed a better identifier.